File: raspberry_pi/rayglides_protocol.py
"""
RayGlides Protocol - Raspberry Pi host-side library.

Mirrors the on-wire format defined in the ESP32 firmware
(src/protocol/RayGlidesProtocol.cpp / .h) byte-for-byte:

    [FRAME_START][msgId][len][payload...][checksum][FRAME_END]

    FRAME_START = 0xAA          FRAME_END = 0x55
    checksum    = msgId ^ len ^ payload[0] ^ payload[1] ^ ... (XOR)

This library is transport-agnostic at its core (encode/decode are pure
functions of bytes) with a thin pyserial wrapper (RayGlidesLink) for the
real use case: talking to the EMS over USB Serial at /dev/ttyACM0.

Three layers, matching how the firmware itself is structured:
  1. encode_frame() / compute_checksum()      - build a valid frame
  2. FrameParser                              - resilient streaming decode,
                                                 tolerant of noise/resync,
                                                 mirrors receiveFrame()'s
                                                 byte-at-a-time behavior
  3. decode_single_frame()                    - strict one-shot decode of
                                                 an exact frame, raises on
                                                 any malformation - used
                                                 for validation/tests
  4. decode_frame() + decode_status_update() / decode_fault_report() /
     decode_ota_ack()                         - turn a Frame's raw
                                                 payload into a plain dict
"""

from dataclasses import dataclass
from typing import Iterator, List, Optional
import time
import threading

try:
    import serial  # pyserial - only required if you actually use RayGlidesLink
except ImportError:  # pragma: no cover - protocol-only usage doesn't need it
    serial = None


# --- Wire-format constants (must match config.h / RayGlidesProtocol.h) ---

FRAME_START = 0xAA
FRAME_END = 0x55
MAX_PAYLOAD = 32

MSG_STATUS_UPDATE = 0x01
MSG_FAULT_REPORT = 0x02
MSG_ACK = 0x03   # Generic transport ACK - non-OTA frames only, see NackReason below
MSG_NACK = 0x04  # Generic transport NACK - non-OTA frames only
MSG_OTA_BEGIN = 0x10
MSG_OTA_DATA = 0x11
MSG_OTA_END = 0x12
MSG_OTA_ABORT = 0x13
MSG_OTA_ACK = 0x14

MSG_NAMES = {
    MSG_STATUS_UPDATE: "STATUS_UPDATE",
    MSG_FAULT_REPORT: "FAULT_REPORT",
    MSG_ACK: "ACK",
    MSG_NACK: "NACK",
    MSG_OTA_BEGIN: "OTA_BEGIN",
    MSG_OTA_DATA: "OTA_DATA",
    MSG_OTA_END: "OTA_END",
    MSG_OTA_ABORT: "OTA_ABORT",
    MSG_OTA_ACK: "OTA_ACK",
}

# Reasons a generic MSG_NACK was sent, carried in payload[0] - must match
# the NackReason enum in RayGlidesProtocol.h exactly. Distinct from
# OTAAckStatus, which covers the separate OTA_BEGIN/DATA/END handshake.
NACK_CHECKSUM_ERROR = 1
NACK_MALFORMED_FRAME = 2
NACK_UNKNOWN_MSG_ID = 3

NACK_REASON_NAMES = {
    NACK_CHECKSUM_ERROR: "CHECKSUM_ERROR",
    NACK_MALFORMED_FRAME: "MALFORMED_FRAME",
    NACK_UNKNOWN_MSG_ID: "UNKNOWN_MSG_ID",
}


# --- Errors ---

class FrameParseError(Exception):
    """A frame was malformed (bad length, missing FRAME_END, too short)."""


class ChecksumError(FrameParseError):
    """A structurally valid frame failed its checksum."""


class AckTimeoutError(Exception):
    """RayGlidesLink.send_and_await() got no ACK/NACK within the timeout,
    across all retries."""


class NackReceived(Exception):
    """RayGlidesLink.send_and_await() got an explicit MSG_NACK back - the
    device understood the frame and rejected it. Not retried automatically:
    resending identical bytes after a checksum/malformed/unknown-msgId
    rejection would just fail the same way again."""

    def __init__(self, reason: int, reason_name: str, acked_msg_id: Optional[int]):
        self.reason = reason
        self.reason_name = reason_name
        self.acked_msg_id = acked_msg_id
        super().__init__(
            f"Device NACKed msg_id={acked_msg_id}: {reason_name} ({reason})")


# --- Encoding ---

def compute_checksum(msg_id: int, length: int, payload: bytes) -> int:
    """Same XOR checksum as computeChecksum() in RayGlidesProtocol.cpp."""
    cs = (msg_id ^ length) & 0xFF
    for b in payload:
        cs ^= b
    return cs & 0xFF


def encode_frame(msg_id: int, payload: bytes = b"") -> bytes:
    """Builds one complete, ready-to-transmit binary frame."""
    if len(payload) > MAX_PAYLOAD:
        raise ValueError(f"payload too long: {len(payload)} > MAX_PAYLOAD={MAX_PAYLOAD}")
    length = len(payload)
    checksum = compute_checksum(msg_id, length, payload)
    return bytes([FRAME_START, msg_id, length]) + bytes(payload) + bytes([checksum, FRAME_END])


# --- Decoding: data model ---

@dataclass
class Frame:
    msg_id: int
    payload: bytes

    @property
    def msg_name(self) -> str:
        return MSG_NAMES.get(self.msg_id, f"UNKNOWN(0x{self.msg_id:02X})")


# --- Decoding: strict, single-frame (validation / tests) ---

def decode_single_frame(raw: bytes) -> Frame:
    """
    Decodes exactly one complete frame from `raw`, which must contain
    nothing but that frame (no leading/trailing bytes). Raises
    FrameParseError / ChecksumError on any malformation. Use this for
    validating a byte sequence you already believe is one whole frame -
    for streaming serial data, use FrameParser instead.
    """
    if len(raw) < 5:
        raise FrameParseError(f"frame too short: {len(raw)} bytes (minimum 5)")
    if raw[0] != FRAME_START:
        raise FrameParseError(f"missing FRAME_START: got 0x{raw[0]:02X}")

    msg_id = raw[1]
    length = raw[2]
    if length > MAX_PAYLOAD:
        raise FrameParseError(f"length {length} exceeds MAX_PAYLOAD={MAX_PAYLOAD}")

    expected_total = 3 + length + 2
    if len(raw) != expected_total:
        raise FrameParseError(
            f"length mismatch: header says {length} payload bytes, "
            f"frame is {len(raw)} bytes (expected {expected_total})")

    payload = bytes(raw[3:3 + length])
    checksum = raw[3 + length]
    end_byte = raw[3 + length + 1]

    if end_byte != FRAME_END:
        raise FrameParseError(f"missing FRAME_END: got 0x{end_byte:02X}")

    expected_checksum = compute_checksum(msg_id, length, payload)
    if checksum != expected_checksum:
        raise ChecksumError(f"checksum mismatch: got 0x{checksum:02X}, expected 0x{expected_checksum:02X}")

    return Frame(msg_id=msg_id, payload=payload)


# --- Decoding: resilient streaming parser (live serial data) ---

class FrameParser:
    """
    Incremental, byte-at-a-time parser for feeding directly from a live
    serial read. Mirrors receiveFrame()'s own tolerance: any byte that
    isn't a valid FRAME_START while idle is silently discarded and
    parsing resyncs on the next byte, exactly like the firmware. A
    structurally valid frame with a bad checksum is dropped (counted,
    not raised) so one corrupted frame never stops the stream - use
    decode_single_frame() instead if you want a hard failure on bad data.
    """

    _WAIT_START, _WAIT_ID, _WAIT_LEN, _WAIT_PAYLOAD, _WAIT_CHECKSUM, _WAIT_END = range(6)

    def __init__(self):
        self.checksum_errors = 0
        self.malformed_frames = 0
        self._log_buffer = bytearray()
        self._reset()

    def _reset(self):
        self._state = self._WAIT_START
        self._msg_id = 0
        self._length = 0
        self._payload = bytearray()
        self._checksum_byte = 0

    def feed(self, data: bytes) -> Iterator[Frame]:
        """Feed raw bytes; yields a Frame for each complete, valid frame found."""
        for byte in data:
            frame = self._feed_byte(byte)
            if frame is not None:
                yield frame

    def _feed_byte(self, byte: int) -> Optional[Frame]:
        if self._state == self._WAIT_START:
            if byte == FRAME_START:
                self._state = self._WAIT_ID
            else:
                if 32 <= byte <= 126 or byte in (10, 13, 9):
                    if byte == 10:  # '\n'
                        try:
                            text = self._log_buffer.decode('utf-8', errors='ignore').strip()
                            if text:
                                print(f"[ESP32 Console] {text}")
                        except Exception:
                            pass
                        self._log_buffer.clear()
                    elif byte != 13:  # skip '\r'
                        self._log_buffer.append(byte)
                else:
                    self._log_buffer.clear()
            return None

        if self._state == self._WAIT_ID:
            self._msg_id = byte
            self._state = self._WAIT_LEN
            return None

        if self._state == self._WAIT_LEN:
            if byte > MAX_PAYLOAD:
                self.malformed_frames += 1
                self._reset()  # Reject and resync, matches firmware
                return None
            self._length = byte
            self._payload = bytearray()
            self._state = self._WAIT_PAYLOAD if self._length > 0 else self._WAIT_CHECKSUM
            return None

        if self._state == self._WAIT_PAYLOAD:
            self._payload.append(byte)
            if len(self._payload) >= self._length:
                self._state = self._WAIT_CHECKSUM
            return None

        if self._state == self._WAIT_CHECKSUM:
            self._checksum_byte = byte
            self._state = self._WAIT_END
            return None

        if self._state == self._WAIT_END:
            msg_id, length, payload, checksum = self._msg_id, self._length, bytes(self._payload), self._checksum_byte
            end_byte = byte
            self._reset()

            if end_byte != FRAME_END:
                self.malformed_frames += 1
                return None

            expected = compute_checksum(msg_id, length, payload)
            if checksum != expected:
                self.checksum_errors += 1
                return None

            return Frame(msg_id=msg_id, payload=payload)

        return None  # pragma: no cover - unreachable


# --- Decoding: known message types -> plain dicts ---

def decode_status_update(payload: bytes) -> dict:
    """Payload: [SOC, ChargeState, BattV, BattI(signed), BattT, SolarV, SolarPower, BattSOH, PrimaryFaultCode]"""
    if len(payload) < 2:
        raise ValueError(f"StatusUpdate payload too short: {len(payload)} bytes")
    result = {"soc": payload[0], "state": payload[1]}
    if len(payload) >= 8:
        current = payload[3]
        result.update({
            "battery_voltage": payload[2],
            "battery_current": current - 256 if current > 127 else current,  # int8_t
            "battery_temp": payload[4],
            "solar_voltage": payload[5],
            "solar_power": payload[6],
            "battery_soh": payload[7],
        })
    if len(payload) >= 9:
        result["active_fault_code"] = payload[8]
    return result


def decode_fault_report(payload: bytes) -> dict:
    """Payload: [FaultCode, Severity]"""
    if len(payload) < 2:
        raise ValueError(f"FaultReport payload too short: {len(payload)} bytes")
    return {"fault_code": payload[0], "severity": payload[1]}


def decode_ota_ack(payload: bytes) -> dict:
    """Payload: [OTAAckStatus, seqLow, seqHigh]"""
    if len(payload) < 3:
        raise ValueError(f"OTA ACK payload too short: {len(payload)} bytes")
    seq = payload[1] | (payload[2] << 8)
    return {"status": payload[0], "seq": seq}


def decode_ack(payload: bytes) -> dict:
    """Payload: [ackedMsgId]"""
    if len(payload) < 1:
        raise ValueError(f"ACK payload too short: {len(payload)} bytes")
    return {"acked_msg_id": payload[0]}


def decode_nack(payload: bytes) -> dict:
    """Payload: [reason, ackedMsgId]"""
    if len(payload) < 1:
        raise ValueError(f"NACK payload too short: {len(payload)} bytes")
    reason = payload[0]
    acked_msg_id = payload[1] if len(payload) >= 2 else None
    return {
        "reason": reason,
        "reason_name": NACK_REASON_NAMES.get(reason, f"UNKNOWN({reason})"),
        "acked_msg_id": acked_msg_id,
    }


def decode_frame(frame: Frame) -> dict:
    """Dispatches a Frame to the right decoder by msg_id. Unknown message
    IDs pass the raw payload through untouched rather than raising, so a
    firmware update that adds a new message type doesn't break older
    host tooling parsing everything else."""
    if frame.msg_id == MSG_STATUS_UPDATE:
        return {"type": "status_update", **decode_status_update(frame.payload)}
    if frame.msg_id == MSG_FAULT_REPORT:
        return {"type": "fault_report", **decode_fault_report(frame.payload)}
    if frame.msg_id == MSG_ACK:
        return {"type": "ack", **decode_ack(frame.payload)}
    if frame.msg_id == MSG_NACK:
        return {"type": "nack", **decode_nack(frame.payload)}
    if frame.msg_id == MSG_OTA_ACK:
        return {"type": "ota_ack", **decode_ota_ack(frame.payload)}
    return {"type": "unknown", "msg_id": frame.msg_id, "payload": bytes(frame.payload)}


# --- Serial link wrapper ---

class RayGlidesLink:
    """
    Thin pyserial wrapper for talking to the RayGlides EMS over USB
    Serial. Defaults match the firmware: /dev/ttyACM0 at 115200 baud.
    Sends real binary frames (encode_frame()) and decodes incoming bytes
    through a FrameParser - this only works correctly against the
    binary-framing version of sendFrame() (see RayGlidesProtocol.cpp);
    it will not parse the older HEX-text debug output.
    """

    def __init__(self, port: str = "/dev/ttyACM0", baudrate: int = 115200, timeout: float = 0.1):
        if serial is None:
            raise RuntimeError("pyserial is required for RayGlidesLink - pip install pyserial")
        self._serial = serial.Serial(port, baudrate, timeout=timeout)
        self._serial_lock = threading.Lock()
        self._parser = FrameParser()
        # Frames read by read_frame() beyond the one it returns (poll()
        # can decode several frames from a single burst of serial data) -
        # queued here instead of dropped, so a fast run of replies
        # (e.g. noise immediately followed by the ACK it's waiting on,
        # both arriving in the same read()) isn't silently lost.
        self._read_queue: List[Frame] = []
        self._read_queue_lock = threading.Lock()
        
        # Thread-safe synchronization dictionary for serial command responses
        self._pending_lock = threading.Lock()
        self._pending_commands = {}  # msg_id -> {"event": threading.Event(), "response": None}

    def close(self):
        with self._serial_lock:
            self._serial.close()

    def __enter__(self):
        return self

    def __exit__(self, *exc_info):
        self.close()

    def send(self, msg_id: int, payload: bytes = b"") -> None:
        with self._serial_lock:
            self._serial.write(encode_frame(msg_id, payload))

    def poll(self) -> List[Frame]:
        """Reads whatever is currently buffered and returns any complete,
        valid frames found (usually 0 or 1). Non-blocking.
        Intercepts ACK and NACK frames to coordinate with waiting send_and_await calls."""
        with self._serial_lock:
            n = self._serial.in_waiting
            if n == 0:
                return []
            data = self._serial.read(n)
            frames = list(self._parser.feed(data))
        
        non_ack_frames = []
        for frame in frames:
            if frame.msg_id in (MSG_ACK, MSG_NACK):
                # ACK payload[0] is the acknowledged msg_id
                # NACK payload[1] is the acknowledged msg_id
                acked_msg_id = None
                if frame.msg_id == MSG_ACK and len(frame.payload) >= 1:
                    acked_msg_id = frame.payload[0]
                elif frame.msg_id == MSG_NACK and len(frame.payload) >= 2:
                    acked_msg_id = frame.payload[1]
                
                if acked_msg_id is not None:
                    with self._pending_lock:
                        if acked_msg_id in self._pending_commands:
                            self._pending_commands[acked_msg_id]["response"] = frame
                            self._pending_commands[acked_msg_id]["event"].set()
                            continue  # Intercepted!
            non_ack_frames.append(frame)
            
        return non_ack_frames

    def read_frame(self, timeout: float = 2.0) -> Optional[Frame]:
        """Blocks up to `timeout` seconds for exactly one frame - handy
        for request/response flows like waiting on an OTA_ACK. Frames
        beyond the one returned (poll() can decode several from a single
        burst) are queued and returned first-in-first-out by later
        read_frame() calls, rather than dropped."""
        with self._read_queue_lock:
            if self._read_queue:
                return self._read_queue.pop(0)
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            frames = self.poll()
            with self._read_queue_lock:
                if frames:
                    self._read_queue.extend(frames[1:])
                    return frames[0]
            time.sleep(0.01)
        return None

    def send_and_await(self, msg_id: int, payload: bytes = b"",
                        timeout: float = 2.0, retries: int = 2) -> Frame:
        """Sends a frame and blocks for the device's MSG_ACK/MSG_NACK reply
        using a thread-safe event listener."""
        event = threading.Event()
        wait_entry = {"event": event, "response": None}
        
        with self._pending_lock:
            self._pending_commands[msg_id] = wait_entry
            
        try:
            for attempt in range(retries + 1):
                event.clear()
                wait_entry["response"] = None
                
                self.send(msg_id, payload)
                
                deadline = time.monotonic() + timeout
                while time.monotonic() < deadline:
                    if event.wait(timeout=0.01):
                        break
                    self.poll()
                
                if event.is_set():
                    resp_frame = wait_entry["response"]
                    if resp_frame is not None:
                        if resp_frame.msg_id == MSG_NACK:
                            decoded = decode_nack(resp_frame.payload)
                            raise NackReceived(decoded["reason"], decoded["reason_name"],
                                                decoded["acked_msg_id"])
                        if resp_frame.msg_id == MSG_ACK:
                            return resp_frame
            raise AckTimeoutError(
                f"No ACK/NACK for msg_id=0x{msg_id:02X} after {retries + 1} attempt(s)")
        finally:
            with self._pending_lock:
                self._pending_commands.pop(msg_id, None)

    @property
    def checksum_errors(self) -> int:
        return self._parser.checksum_errors

    @property
    def malformed_frames(self) -> int:
        return self._parser.malformed_frames

File: RayGlides_EMS/src/protocol/RayGlidesProtocol.cpp
#include "RayGlidesProtocol.h"
#include "config.h"
#include "../debug/DebugLog.h"
#include "../ota/OTAUpdate.h"  // isOTAInProgress() - suppresses generic NACKs mid-transfer
#include "sensors/SensorSimulator.h"
#include "configuration/ConfigManager.h"
#include "charging/ChargingDecision.h"
#include <Preferences.h>

static unsigned long lastMessageTimestamp = 0;

// --- Sending ---

uint8_t computeChecksum(uint8_t msgId, uint8_t len, uint8_t* payload) {
  if (len > MAX_PAYLOAD) len = MAX_PAYLOAD;
  uint8_t cs = msgId ^ len;
  for (int i = 0; i < len; i++) cs ^= payload[i];
  return cs;
}

void sendFrame(uint8_t msgId, uint8_t* payload, uint8_t len) {
  if (len > MAX_PAYLOAD) len = MAX_PAYLOAD;
  uint8_t checksum = computeChecksum(msgId, len, payload);


  // Real binary wire format - START, msgId, len, payload..., checksum, END -
  // written as raw bytes via Serial.write(), NOT printed as ASCII/HEX text.
  // This must exactly match what receiveFrame() below (and the Raspberry
  // Pi FrameParser) parse byte-for-byte.
  uint8_t frame[3 + MAX_PAYLOAD + 2];
  uint8_t idx = 0;
  frame[idx++] = FRAME_START;
  frame[idx++] = msgId;
  frame[idx++] = len;
  for (int i = 0; i < len; i++) frame[idx++] = payload[i];
  frame[idx++] = checksum;
  frame[idx++] = FRAME_END;
  Serial.write(frame, idx);

  // Human-readable trace goes through the debug logger, not the wire -
  // this keeps binary framing and text diagnostics on genuinely separate
  // channels (logInfo() also auto-silences during an active OTA transfer,
  // which matters since OTA reuses this same sendFrame() for its ACKs).
  char msg[32];
  snprintf(msg, sizeof(msg), "TX msgId=0x%02X len=%d", msgId, len);
  logInfo("PROTO", msg);
}

void sendStatusUpdate(BatteryData battery, SolarData solar, ChargeState state) {
  // Payload: [SOC, ChargeState, BattV, BattI(signed), BattT, SolarV, SolarPower, BattSOH, PrimaryFaultCode]
  uint8_t payload[9] = {
    (uint8_t)battery.soc,
    (uint8_t)state,
    (uint8_t)constrain((int)round(battery.voltage), 0, 255),
    (uint8_t)(int8_t)constrain((int)round(battery.current), -128, 127),
    (uint8_t)constrain((int)round(battery.temperature), 0, 255),
    (uint8_t)constrain((int)round(solar.voltage), 0, 255),
    (uint8_t)constrain((int)round(solar.power), 0, 255),
    (uint8_t)constrain(battery.soh, 0, 255),
    (uint8_t)currentPrimaryFault
  };
  sendFrame(MSG_STATUS_UPDATE, payload, 9);
}

void sendFaultReport(FaultCode code, Severity sev) {
  uint8_t payload[2] = { (uint8_t)code, (uint8_t)sev };
  sendFrame(MSG_FAULT_REPORT, payload, 2);
}

void sendAck(uint8_t ackedMsgId) {
  uint8_t payload[1] = { ackedMsgId };
  sendFrame(MSG_ACK, payload, 1);
}

void sendNack(NackReason reason, uint8_t ackedMsgId) {
  uint8_t payload[2] = { (uint8_t)reason, ackedMsgId };
  sendFrame(MSG_NACK, payload, 2);
}

// Shared by every rejection path in receiveFrame(): suppresses the NACK
// during an active OTA transfer (see header doc) and keeps the log
// message + wire NACK consistent in one place instead of duplicated at
// each call site.
static void rejectFrame(const char* logMsg, NackReason reason, uint8_t msgId) {
  logWarn("PROTO", logMsg);
  if (!isOTAInProgress()) {
    sendNack(reason, msgId);
  }
}

// --- Receiving ---
// Non-blocking: only attempts a read if a start byte is already waiting.
// Intended for a real second device (dashboard/BMS) transmitting frames
// back to the EMS. In a single-device Wokwi simulation this will simply
// find nothing to read, which is expected.

ReceivedFrame receiveFrame() {
  ReceivedFrame result;
  result.valid = false;
  result.length = 0;

  if (Serial.available() < 3) return result;
  if (Serial.peek() != FRAME_START) {
    Serial.read();  // Discard stray byte, resync on next call
    return result;
  }

  uint8_t start = Serial.read();
  uint8_t msgId = Serial.read();
  uint8_t length = Serial.read();

  if (length > MAX_PAYLOAD) {  // Malformed - reject
    rejectFrame("Malformed frame - length exceeds MAX_PAYLOAD", NACK_MALFORMED_FRAME, msgId);
    return result;
  }

  unsigned long startWait = millis();
  while ((uint8_t)Serial.available() < (uint8_t)(length + 2)) {
    if (millis() - startWait > 10) {  // 10ms timeout
      rejectFrame("Malformed frame - receive timeout", NACK_MALFORMED_FRAME, msgId);
      return result;
    }
    delay(1);
  }

  uint8_t payload[MAX_PAYLOAD];
  for (int i = 0; i < length; i++) payload[i] = Serial.read();
  uint8_t receivedChecksum = Serial.read();
  uint8_t endByte = Serial.read();

  if (endByte != FRAME_END) {  // Malformed frame
    rejectFrame("Malformed frame - missing FRAME_END", NACK_MALFORMED_FRAME, msgId);
    return result;
  }

  uint8_t expectedChecksum = computeChecksum(msgId, length, payload);
  if (receivedChecksum != expectedChecksum) {
    sendFaultReport(F009_CHECKSUM_ERROR, SEV_WARNING);  // Existing safety/fault-log path - unchanged
    rejectFrame("Checksum mismatch on received frame", NACK_CHECKSUM_ERROR, msgId);
    return result;
  }

  result.valid = true;
  result.msgId = msgId;
  result.length = length;
  for (int i = 0; i < length; i++) result.payload[i] = payload[i];
  noteMessageReceived();
  return result;
}

void handleIncomingFrame(ReceivedFrame frame) {
  if (!frame.valid) return;

  char msg[64];
  if (frame.msgId == MSG_STATUS_UPDATE) {
    if (frame.length >= 5) {
      snprintf(msg, sizeof(msg), "RX StatusUpdate SOC=%d State=%d V=%d I=%d T=%d",
        frame.payload[0], frame.payload[1], frame.payload[2],
        (int8_t)frame.payload[3], frame.payload[4]);
    } else {
      snprintf(msg, sizeof(msg), "RX StatusUpdate SOC=%d State=%d", frame.payload[0], frame.payload[1]);
    }
    logInfo("PROTO", msg);
    sendAck(frame.msgId);
  } else if (frame.msgId == MSG_FAULT_REPORT) {
    snprintf(msg, sizeof(msg), "RX FaultReport Code=%d Severity=%d", frame.payload[0], frame.payload[1]);
    logInfo("PROTO", msg);
    if (frame.length >= 1 && frame.payload[0] == 0) {
      clearAllFaults();
    }
    sendAck(frame.msgId);
  } else if (frame.msgId == MSG_SET_SCENARIO) {
    if (frame.length >= 1) {
      setSimulationScenario(frame.payload[0]);
      sendAck(frame.msgId);
    } else {
      sendNack(NACK_MALFORMED_FRAME, frame.msgId);
    }
  } else if (frame.msgId == MSG_SET_CONFIG) {
    if (frame.length >= 5) {
      uint8_t paramId = frame.payload[0];
      float val;
      memcpy(&val, &frame.payload[1], 4);
      if (updateConfigParameter(paramId, val)) {
        sendAck(frame.msgId);
      } else {
        if (paramId >= 1 && paramId <= 8) {
          sendNack(NACK_INVALID_VALUE, frame.msgId);
        } else {
          sendNack(NACK_MALFORMED_FRAME, frame.msgId);
        }
      }
    } else {
      sendNack(NACK_MALFORMED_FRAME, frame.msgId);
    }
  } else if (frame.msgId == MSG_GET_STATUS) {
    extern BatteryData battery;
    extern SolarData solar;
    extern ChargeState currentState;
    sendStatusUpdate(battery, solar, currentState);
    sendAck(frame.msgId);
  } else if (frame.msgId == MSG_SET_MODE) {
    if (frame.length >= 1) {
      uint8_t mVal = frame.payload[0];
      extern ChargingMode manualModeOverride;
      extern bool useManualMode;
      if (mVal == 255) {
        useManualMode = false;
        sendAck(frame.msgId);
      } else if (mVal >= 0 && mVal <= 3) {
        useManualMode = true;
        manualModeOverride = (ChargingMode)mVal;
        sendAck(frame.msgId);
      } else {
        sendNack(NACK_INVALID_VALUE, frame.msgId);
      }
    } else {
      sendNack(NACK_MALFORMED_FRAME, frame.msgId);
    }
  } else if (frame.msgId == MSG_SET_FAN) {
    if (frame.length >= 1) {
      uint8_t fVal = frame.payload[0];
      extern void setFanManualOverride(bool enable, float duty);
      if (fVal == 255) {
        setFanManualOverride(false, 0.0f);
        sendAck(frame.msgId);
      } else if (fVal <= 100) {
        setFanManualOverride(true, fVal / 100.0f);
        sendAck(frame.msgId);
      } else {
        sendNack(NACK_INVALID_VALUE, frame.msgId);
      }
    } else {
      sendNack(NACK_MALFORMED_FRAME, frame.msgId);
    }
  } else if (frame.msgId == MSG_RESTART) {
    sendAck(frame.msgId);
    delay(500);
    ESP.restart();
  } else if (frame.msgId == MSG_SET_BROKER) {
    if (frame.length >= 1) {
      char broker[33];
      uint8_t len = frame.length;
      if (len > 32) len = 32;
      memcpy(broker, frame.payload, len);
      broker[len] = '\0';
      
      Preferences prefs;
      prefs.begin("wifi", false);
      prefs.putString("broker", broker);
      prefs.end();
      
      logInfo("PROTO", ("Set broker IP dynamically to: " + String(broker)).c_str());
      sendAck(frame.msgId);
      
      extern void initMQTT();
      initMQTT();
    } else {
      sendNack(NACK_MALFORMED_FRAME, frame.msgId);
    }
  } else if (frame.msgId == MSG_ACK || frame.msgId == MSG_NACK) {
    // Never ACK/NACK an ACK/NACK - a stray echo of our own reply (or the
    // host's) would otherwise ping-pong indefinitely.
    snprintf(msg, sizeof(msg), "RX %s payload[0]=%d",
      frame.msgId == MSG_ACK ? "Ack" : "Nack", frame.payload[0]);
    logInfo("PROTO", msg);
  } else {
    snprintf(msg, sizeof(msg), "RX Unknown MSG_ID=0x%02X", frame.msgId);
    logWarn("PROTO", msg);
    sendNack(NACK_UNKNOWN_MSG_ID, frame.msgId);
  }
}

// --- Timeout watchdog ---

void noteMessageReceived() {
  lastMessageTimestamp = millis();
}

bool checkCommunicationTimeout() {
  if (millis() - lastMessageTimestamp > COMM_TIMEOUT_MS) {
    return true;   // Caller decides whether to raise F008
  }
  return false;
}
